Add `dict()` constructor

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51934 Add `dict()` constructor**

Differential Revision: [D26418199](https://our.internmc.facebook.com/intern/diff/D26418199)

Commit:
Add `dict()` constructor

[ghstack-poisoned]

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3007,10 +3021,164 @@ struct to_ir {
       }
-      default:
-        TORCH_INTERNAL_ASSERT(false, "unknown special form: ", form);
+      if (!apply.inputs().empty() && apply.inputs()[0].kind() == TK_DICT_LITERAL) {
```
IMO, we should get rid of the special empty handling - we're not handling dicts/mappings in general, and `dict([])` will work without the special case.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3007,10 +3021,164 @@ struct to_ir {
+    NamedValue self = NamedValue(dc.range(), "self", dc_node);
+
+    for (auto kwarg : apply.attributes()) {
+      auto name = StringLiteral::create(kwarg.range(), kwarg.name().name());
+      auto k = emitExpr(name);
+      auto v = emitExpr(kwarg.value());
+      NamedValue input_k = NamedValue(kwarg.range(), "", k);
+      NamedValue input_v = NamedValue(kwarg.range(), "", v);
+    emitBuiltinCall(
+      kwarg.range(), *graph, aten::_set_item, {self, input_k, input_v}, {});
+    }
```
This logic is very similar to 3076-3083, you might be able to factor it out to a helper lambda or function

eellison reviewed (approved):
Looks great 🚀 🚀 🚀

Commit:
Update on "Add `dict()` constructor"




[ghstack-poisoned]

Commit:
Update on "Add `dict()` constructor"


Differential Revision: [D26418199](https://our.internmc.facebook.com/intern/diff/D26418199)

[ghstack-poisoned]

Commit:
Update on "Add `dict()` constructor"


Differential Revision: [D26418199](https://our.internmc.facebook.com/intern/diff/D26418199)

[ghstack-poisoned]
